### ISSUE TITLE
Suppport for MongoDb $setOnInsert

### ISSUE THREAD

**Issue by valeeum** [ZJONSSON/node-etl#44]
Currently, there is no way to add additional commands, such as $setOnInsert, to the Mongo bulk operation.  Use case is to set a 'dateCreated' field when a document is inserted when _upsert_ is set to true.

**Comment by ZJONSSON** [comment]
Thanks @valeeum, this would be a very nice improvement.  Here is a suggested implementation:   If `$setOnInsert` is provided in the `options` of `etl.mongo.update` then:
* if its an Object, then the same Object will be included as `$setOnInsert` for every data packet in the bulk update
* if its a function it will be called with the data for each packet and the results will be included in the corresponding `$setOnInsert`

So if you wanted to set `dateCreated` you could either do: `etl.mongo.update(collection,keys,{upsert:true, $setOnInsert: new Date()})`   (which will set all inserted records to the date when the bulk started) or `etl.mongo.update(collection,keys,{upsert:true, $setOnInsert: function() { return new Date();}})` which would set it to the specific time each record was included in a bulk.

Let me know what you think or if you have other ideas?

**Comment by valeeum** [comment]
sorry about the delayed response...i was out of town.  

i love the current minimalistic api but would like to have the additional flexibility to completely modify the argument passed in to the _op.updateOne_ function. 

`op.updateOne({$set:d});`

If the callback were supplied with the _d_ variable as an argument, that would give the user the flexibility to craft whatever update operation they would like.  Field update operators are provided here.  As you can see, _$set_ is just one of many operators available to us.  

https://docs.mongodb.com/manual/reference/operator/update-field.

**Comment by ZJONSSON** [comment]
Ah that makes sense - how about adding a new generic function `etl.mongo.bulk` that simply passes on each incoming object directly to `.updateOne` without wrapping it inside `$set`.   The keys of each objects would have to be any of the operators listed in the reference above?   Let me know if you have better ideas?

**Comment by valeeum** [comment]
That actually sounds great!

**Comment by valeeum** [comment]
Take a look at the changes I made here...

https://github.com/vcomedia/node-etl/blob/master/lib/mongo/update.js

I actually modified the existing etl.mongo.update function to allow flexibility to accommodate this use case as well as allow for backward compatibility (backward compatibility not yet tested but theoretically should work). 

So we can now support feeding the following object to the function....

```
{
	$setOnInsert: {
		dateCreated: new Date(),
	},
	$set: {
		dateModified: new Date()
	},
	$push: {
		someArray: {
                   id: 'someID',
                   type: 'someType'
                }
	}
}
```

with the following keys...

['$push.someArray.id','$push.someArray.type']

The function will now walk the object using the dotted string notation and discard the first component starting with '$' character when creating the criteria since that signifies an _update operation_.  So in this case, the criteria becomes: 

```
{
	someArray.id : 'someID',
        someArray.type : 'someType',
}
```

You can still use objects that represent property + values like the previous version as well as keys that don't include update operators.  Only difference is that now there is support for dot notation for the query.

Let me know what you think.

**Comment by ZJONSSON** [comment]
Sorry for the delay - I'm back at my desk after the holidays and will take a better look in the next couple of days.  Happy new year!

**Comment by valeeum** [comment]
I want to note I made one more update to address an error in mongo 3.4 caused by passing in the entire options variable into the bulk execute method.  The new format I propose is to explicitly set the writeConcern object.  This of course will break backward compatibility.  

```
      var writeConcern = self.options.writeConcern  ? self.options.writeConcern  : null;
      bulk.execute(writeConcern,function(err,d) {
        cb(err,self.options.pushResult && d);
      });
```

Spec is available here:

https://docs.mongodb.com/manual/reference/method/Bulk.execute/

The entire update is available here:

https://github.com/vcomedia/node-etl/blob/master/lib/mongo/update.js

**Comment by ZJONSSON** [comment]
Thanks again @valeeum - I folded your code into `etl.mongo.bulk` to make it available for use.
I was a little hesitant to change the `update` as this would break backwards compatibility.
We need to add tests and docs, and when we do a major release we could merge with update

published as `+ etl@0.3.38`
